File: files/CSEC 07.07.R
library(tidyverse)
library(readxl)

path = "files/Excel Challenge  7th July.xlsx"

input = read_xlsx(path, range = "B3:B20", col_types = "date")
test  = read_xlsx(path, range = "D3:D8")

result = input %>%
  filter((is.na(lag(SALES)) | is.na(lead(SALES))) & (!row_number() %in% c(1, n())))

identical(result$SALES, test$`Before & After Dates`)
#> [1] TRUE

File: files/2024-07-07/CSEC 07.07.py
import pandas as pd
import numpy as np

path = "files/Excel Challenge  7th July.xlsx"

input = pd.read_excel(path, usecols="B", skiprows = 2, nrows = 17)
test  = pd.read_excel(path, usecols="D", skiprows = 2, nrows = 5)

result = np.where((input["SALES"].shift() == "NEW ORDER") | (input["SALES"].shift(-1) == "NEW ORDER"), 1, 0)
result = input[result == 1].reset_index(drop=True)
result["SALES"] = result["SALES"].astype("datetime64[ns]")

print(result["SALES"].equals(test["Before & After Dates"])) # True